Blu magic core updates

MowFord opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

This ports the SATA code from CEXI for blu magic.

It also enables TA to function for blu magic skillchain damage

End result is (using cannonball as an example)
- a SA cannonball will auto crit
- if taChar is in position
  - a TA cannonball will transfer enmity
  - a TA cannonball that skillchains will transfer enmity of both spell and skillchain

`PHYSICAL_POTENCY` now gives acc and bonus to cratio

Finally, some adjustments were made to support `params.bonusacc` and `params.critchance`. If these params aren't passed both will be zero by default

_Also_ (sorry for the late discovery, but didn't want to make a new PR since this one is still open): Found that the base multiplier is being set to 1 unless Chain Affinity is in effect. This is wrong: https://www.bg-wiki.com/ffxi/Calculating_Blue_Magic_Damage

The base multiplier is in the table as ftp0 and is only applied to the first hit.

## Steps to test these changes

`!stun` and `!immortal` a mob
position for SA and compare damage before and after cannonball
position for TA and remove `!stun` after to see the mob target the player
TA for skillchain damage is a bit trickier to test, but you can use `!getenmity` to confirm your enmity doesn't spike after using TA (open the skillchain with the second player)

kelppo committed:
Implement Sneak Attack interaction with physical Blue Magic

kelppo committed:
Implement Trick Attack interaction with physical Blue Magic

MowFord commented:
This is an initial submission, I'm going to confirm I can do a removal of the `, nil)` addition to a few spells as well as fully test it

Xaver-DaRed commented on scripts/globals/bluemagic.lua:
```diff
@@ -220,23 +220,36 @@ xi.spells.blue.usePhysicalSpell = function(caster, target, spell, params)
-                finaldmg = finaldmg + (finalD * (multiplier + correlationMultiplier) * pdif) -- first hit gets full multiplier
+            if hitsdone == 0 then -- first hit gets full multiplier
+                if sneakIsApplicable then
+                    finaldmg = finaldmg + (finalD * (multiplier + correlationMultiplier) * (pdif + 0.7))
```
finaldmg = finaldmg + finalD * (multiplier + correlationMultiplier) * (pdif + 0.7)

(Unneeded parenthesis)

Xaver-DaRed commented on scripts/globals/bluemagic.lua:
```diff
@@ -220,23 +220,36 @@ xi.spells.blue.usePhysicalSpell = function(caster, target, spell, params)
+                if sneakIsApplicable then
+                    finaldmg = finaldmg + (finalD * (multiplier + correlationMultiplier) * (pdif + 0.7))
+                else
+                    finaldmg = finaldmg + (finalD * (multiplier + correlationMultiplier) * pdif)
```
finaldmg = finaldmg + finalD * (multiplier + correlationMultiplier) * pdif

(Unneeded parenthesis)

WinterSolstice8 commented on scripts/globals/bluemagic.lua:
```diff
@@ -46,6 +46,29 @@ local function calculateWSC(attacker, params)
     return wsc
 end
 
+local function calculateDEXvsAGICritRate(attacker, target)
```
Should move this into https://github.com/LandSandBoat/server/blob/base/scripts/globals/combat/physical_utilities.lua

Xaver-DaRed replied:
IF thats the case, I have a request:
```lua
xi.combat.physical.calculateMeleeCriticalRate = function(actor, target)
    -- See reference at https://www.bg-wiki.com/ffxi/Critical_Hit_Rate
    local criticalRate = 0
    local dDex         = actor:getStat(xi.mod.DEX) - target:getStat(xi.mod.AGI)

    if dDex > 50 then
        criticalRate = 0.15
    elseif dDex >= 40 then
        criticalRate = (dDex - 35) / 100
    elseif dDex >= 30 then
        criticalRate = 0.04
    elseif dDex >= 20 then
        criticalRate = 0.03
    elseif dDex >= 14 then
        criticalRate = 0.02
    elseif dDex >= 7 then
        criticalRate = 0.01
    end

    -- We should seriously consider adding merits, job points and other modifiers here. 
    
    return criticalRate
end
```

Reasoning:
- This files are the base of weaponskills + other actions being rewriten to be... less chaotic.
- Variable naming: I personally prefer to name variables after what wikis name them. Same terms = more clarity. I also dislike abreviations.
- Order: Critical hit rate is already defined as 0, so theres no need for a step setting it to 0.
- We should also really consider puting mods that affect critical rate in there aswell. Because theres no need to add them in a different function, nor independently in blue magic, weaponskills, TP moves, JAs, etc

Xaver-DaRed replied:
Those are my dominions. Feel spooked. UHHHhHhHHhH.... :ghost:

MowFord replied:
Done and done.

Added a third param that if passed as true, bypasses the other modifiers so i could change weaponskills.lua as little as possible (and for flexibility)

WinterSolstice8 commented on scripts/globals/combat/physical_utilities.lua:
```diff
@@ -428,6 +428,34 @@ xi.combat.physical.calculateMeleePDIF = function(actor, target, weaponType, wsAt
+-- calculates melee critical rate, if true is passed, then only returns crit from dex vs agi
+xi.combat.physical.calculateMeleeCriticalRate = function(actor, target, nativeOnly)
+    -- See reference at https://www.bg-wiki.com/ffxi/Critical_Hit_Rate
+    local nativecrit = 0
```
Shouldn't this be 0.05? That's the base crit rate -- the rest should be adding to it, no?

Xaver-DaRed replied:
The elseif chain below could also use some simplification.

```lua
xi.combat.physical.calculateMeleeCriticalRate = function(actor, target)
    -- See reference at https://www.bg-wiki.com/ffxi/Critical_Hit_Rate
    local finalCriticalRate = 0
    local baseCriticalRate  = 0.05
    local dDex              = actor:getStat(xi.mod.DEX) - target:getStat(xi.mod.AGI)
    local statBonus         = 0

    if dDex > 50 then
        statBonus = 0.15
    elseif dDex >= 40 then
        statBonus = (dDex - 35) / 100
    elseif dDex >= 30 then
        statBonus = 0.04
    elseif dDex >= 20 then
        criticalRate = 0.03
    elseif dDex >= 14 then
        statBonus = 0.02
    elseif dDex >= 7 then
        statBonus = 0.01
    end
    
    finalCriticalRate = baseCriticalRate + statBonus + (actor:getMod(xi.mod.CRITHITRATE) + (actor:getMerit(xi.merit.CRIT_HIT_RATE) - target:getMerit(xi.merit.ENEMY_CRIT_RATE)) / 100
    
    return finalCriticalRate
end
```

I also dont like that `nativeOnly` argument in there.
In which cases would we want the critical rate from dStat only?

MowFord replied:
Yea the goal of this function was to be a drop-in replacement for what was used in `weaponskills.lua` as well as for use in `bluemagic.lua`. 

You might be right that we should include the base .05, i just copied the function over. Ditto for the chained if/else.

WinterSolstice8 replied:
On second thought, go with Xaver's function here

MowFord replied:
Understood. I've tweaked as I think youl would want while keeping the functions separate for base and extra mods

Xaver-DaRed commented on scripts/globals/combat/physical_utilities.lua:
```diff
@@ -428,6 +428,46 @@ xi.combat.physical.calculateMeleePDIF = function(actor, target, weaponType, wsAt
+    elseif dDex >= 30 then
+        statBonus = 0.04
+    elseif dDex >= 20 then
+        criticalRate = 0.03
```
error

MowFord replied:
![image](https://github.com/LandSandBoat/server/assets/131182600/ac9846c3-e5a0-4967-a474-9e907aff5ef2)

I'm never copying from your test again

Xaver-DaRed replied:
I never said I tested it

MowFord replied:
Btw if you squash we lose the co-author commits i cherry-picked

TeoTwawki replied:
WELL THEN! LETS LEARN ADVANCED REBASING! (by "lets" I mean "you people who are not me")

TeoTwawki replied:
...I ***guess*** you could just put all the people as "co authored by" in the new single commit but that's a lot less ~~fun for me to watch~~ _learning a valuable new skill._

Xaver-DaRed replied:
I actually meant to just `hold` it. Couse now I need to see what we have in weaponskills.lua

Couse I'll be honest, I dont see why we cant just have all logic regarding critical hit rate in 1 function instead of several, even if that means purging whats already in weaponskills.lua.

The only thing that could potentially conflict is the fTP step. Maybe. A mere matter of just passing nill and doing a nill check, maybe. I dunnow, hence why I have to read.

MowFord replied:
Well this PR does remove the crit rate function in ws.lua 

I made two functions because you didn't like the parameter in a single one

And by "you" I mean "you people who are not me"

MowFord committed:
Enable TA for phys blu skillchain dmg

Install plumbing for "varies with TP" wrt acc and crit
Moves melee crit calculation function into phs_utils
base multiplier inherits param where defined (only applied to first hit)
